File: binding/src/sql/libqsqlquery.cpp
#include <QAnyStringView>
#include <QSqlDatabase>
#include <QSqlDriver>
#include <QSqlError>
#include <QSqlQuery>
#include <QSqlRecord>
#include <QSqlResult>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QVariant>
#include <qsqlquery.h>
#include "libqsqlquery.h"
#include "libqsqlquery.hxx"

QSqlQuery* QSqlQuery_new(QSqlResult* r) {
	 return new QSqlQuery(r);
}

QSqlQuery* QSqlQuery_new2() {
	 return new QSqlQuery();
}

QSqlQuery* QSqlQuery_new3(const QSqlDatabase* db) {
	 return new QSqlQuery(*db);
}

QSqlQuery* QSqlQuery_new4(const QSqlQuery* other) {
	 return new QSqlQuery(*other);
}

QSqlQuery* QSqlQuery_new5(const libqt_string query) {
	 return new QSqlQuery(QString::fromUtf8(query.data, query.len));
}

QSqlQuery* QSqlQuery_new6(const libqt_string query, const QSqlDatabase* db) {
	 return new QSqlQuery(QString::fromUtf8(query.data, query.len), *db);
}

void QSqlQuery_OperatorAssign(QSqlQuery* self, const QSqlQuery* other) {
	self->operator=(*other);
}

void QSqlQuery_Swap(QSqlQuery* self, QSqlQuery* other) {
	self->swap(*other);
}

bool QSqlQuery_IsValid(const QSqlQuery* self) {
	return self->isValid();
}

bool QSqlQuery_IsActive(const QSqlQuery* self) {
	return self->isActive();
}

bool QSqlQuery_IsNull(const QSqlQuery* self, int field) {
	return self->isNull(field);
}

bool QSqlQuery_IsNull2(const QSqlQuery* self, libqt_string name) {
	return self->isNull(QAnyStringView(QString::fromUtf8(name.data, name.len)));
}

int QSqlQuery_At(const QSqlQuery* self) {
	return self->at();
}

libqt_string QSqlQuery_LastQuery(const QSqlQuery* self) {
	QString _ret = self->lastQuery();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

int QSqlQuery_NumRowsAffected(const QSqlQuery* self) {
	return self->numRowsAffected();
}

QSqlError* QSqlQuery_LastError(const QSqlQuery* self) {
	return new QSqlError(self->lastError());
}

bool QSqlQuery_IsSelect(const QSqlQuery* self) {
	return self->isSelect();
}

int QSqlQuery_Size(const QSqlQuery* self) {
	return self->size();
}

const QSqlDriver* QSqlQuery_Driver(const QSqlQuery* self) {
	return self->driver();
}

const QSqlResult* QSqlQuery_Result(const QSqlQuery* self) {
	return self->result();
}

bool QSqlQuery_IsForwardOnly(const QSqlQuery* self) {
	return self->isForwardOnly();
}

QSqlRecord* QSqlQuery_Record(const QSqlQuery* self) {
	return new QSqlRecord(self->record());
}

void QSqlQuery_SetForwardOnly(QSqlQuery* self, bool forward) {
	self->setForwardOnly(forward);
}

bool QSqlQuery_Exec(QSqlQuery* self, const libqt_string query) {
	return self->exec(QString::fromUtf8(query.data, query.len));
}

QVariant* QSqlQuery_Value(const QSqlQuery* self, int i) {
	return new QVariant(self->value(i));
}

QVariant* QSqlQuery_Value2(const QSqlQuery* self, libqt_string name) {
	return new QVariant(self->value(QAnyStringView(QString::fromUtf8(name.data, name.len))));
}

void QSqlQuery_SetNumericalPrecisionPolicy(QSqlQuery* self, int precisionPolicy) {
	self->setNumericalPrecisionPolicy(static_cast<QSql::NumericalPrecisionPolicy>(precisionPolicy));
}

int QSqlQuery_NumericalPrecisionPolicy(const QSqlQuery* self) {
	return self->numericalPrecisionPolicy();
}

void QSqlQuery_SetPositionalBindingEnabled(QSqlQuery* self, bool enable) {
	self->setPositionalBindingEnabled(enable);
}

bool QSqlQuery_IsPositionalBindingEnabled(const QSqlQuery* self) {
	return self->isPositionalBindingEnabled();
}

bool QSqlQuery_Seek(QSqlQuery* self, int i) {
	return self->seek(i);
}

bool QSqlQuery_Next(QSqlQuery* self) {
	return self->next();
}

bool QSqlQuery_Previous(QSqlQuery* self) {
	return self->previous();
}

bool QSqlQuery_First(QSqlQuery* self) {
	return self->first();
}

bool QSqlQuery_Last(QSqlQuery* self) {
	return self->last();
}

void QSqlQuery_Clear(QSqlQuery* self) {
	self->clear();
}

bool QSqlQuery_Exec2(QSqlQuery* self) {
	return self->exec();
}

bool QSqlQuery_ExecBatch(QSqlQuery* self) {
	return self->execBatch();
}

bool QSqlQuery_Prepare(QSqlQuery* self, const libqt_string query) {
	return self->prepare(QString::fromUtf8(query.data, query.len));
}

void QSqlQuery_BindValue(QSqlQuery* self, const libqt_string placeholder, const QVariant* val) {
	self->bindValue(QString::fromUtf8(placeholder.data, placeholder.len), *val);
}

void QSqlQuery_BindValue2(QSqlQuery* self, int pos, const QVariant* val) {
	self->bindValue(pos, *val);
}

void QSqlQuery_AddBindValue(QSqlQuery* self, const QVariant* val) {
	self->addBindValue(*val);
}

QVariant* QSqlQuery_BoundValue(const QSqlQuery* self, const libqt_string placeholder) {
	return new QVariant(self->boundValue(QString::fromUtf8(placeholder.data, placeholder.len)));
}

QVariant* QSqlQuery_BoundValue2(const QSqlQuery* self, int pos) {
	return new QVariant(self->boundValue(pos));
}

libqt_list QSqlQuery_BoundValues(const QSqlQuery* self) {
	return self->boundValues();
}

libqt_list QSqlQuery_BoundValueNames(const QSqlQuery* self) {
	return self->boundValueNames();
}

libqt_string QSqlQuery_BoundValueName(const QSqlQuery* self, int pos) {
	QString _ret = self->boundValueName(pos);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QSqlQuery_ExecutedQuery(const QSqlQuery* self) {
	QString _ret = self->executedQuery();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QVariant* QSqlQuery_LastInsertId(const QSqlQuery* self) {
	return new QVariant(self->lastInsertId());
}

void QSqlQuery_Finish(QSqlQuery* self) {
	self->finish();
}

bool QSqlQuery_NextResult(QSqlQuery* self) {
	return self->nextResult();
}

bool QSqlQuery_Seek2(QSqlQuery* self, int i, bool relative) {
	return self->seek(i, relative);
}

bool QSqlQuery_ExecBatch1(QSqlQuery* self, int mode) {
	return self->execBatch(static_cast<QSqlQuery::BatchExecutionMode>(mode));
}

void QSqlQuery_BindValue3(QSqlQuery* self, const libqt_string placeholder, const QVariant* val, int typeVal) {
	self->bindValue(QString::fromUtf8(placeholder.data, placeholder.len), *val, static_cast<QFlags<QSql::ParamTypeFlag>>(typeVal));
}

void QSqlQuery_BindValue32(QSqlQuery* self, int pos, const QVariant* val, int typeVal) {
	self->bindValue(pos, *val, static_cast<QFlags<QSql::ParamTypeFlag>>(typeVal));
}

void QSqlQuery_AddBindValue2(QSqlQuery* self, const QVariant* val, int typeVal) {
	self->addBindValue(*val, static_cast<QFlags<QSql::ParamTypeFlag>>(typeVal));
}

void QSqlQuery_Delete(QSqlQuery* self) {
    delete self;
}

File: binding/include/sql/libqsqlquery.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qsqlquery.html
pub const qsqlquery = struct {

    /// New constructs a new QSqlQuery object.
    pub fn New(r: ?*anyopaque) QtC.QSqlQuery {
        return qtc.QSqlQuery_new(@ptrCast(r));
    }


    /// New2 constructs a new QSqlQuery object.
    pub fn New2() QtC.QSqlQuery {
        return qtc.QSqlQuery_new2();
    }


    /// New3 constructs a new QSqlQuery object.
    pub fn New3(db: ?*anyopaque) QtC.QSqlQuery {
        return qtc.QSqlQuery_new3(@ptrCast(db));
    }


    /// New4 constructs a new QSqlQuery object.
    pub fn New4(other: ?*anyopaque) QtC.QSqlQuery {
        return qtc.QSqlQuery_new4(@ptrCast(other));
    }


    /// New5 constructs a new QSqlQuery object.
    pub fn New5(query: []const u8) QtC.QSqlQuery {
        const query_str = qtc.libqt_string{
    .len = query.len,
    .data = query.ptr,
};
return qtc.QSqlQuery_new5(query_str);
    }


    /// New6 constructs a new QSqlQuery object.
    pub fn New6(query: []const u8, db: ?*anyopaque) QtC.QSqlQuery {
        const query_str = qtc.libqt_string{
    .len = query.len,
    .data = query.ptr,
};
return qtc.QSqlQuery_new6(query_str, @ptrCast(db));
    }


    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QSqlQuery_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QSqlQuery_Swap(@ptrCast(self), @ptrCast(other));
    }

    pub fn IsValid(self: ?*anyopaque, ) bool {
        return qtc.QSqlQuery_IsValid(@ptrCast(self));
    }

    pub fn IsActive(self: ?*anyopaque, ) bool {
        return qtc.QSqlQuery_IsActive(@ptrCast(self));
    }

    pub fn IsNull(self: ?*anyopaque, field: i32) bool {
        return qtc.QSqlQuery_IsNull(@ptrCast(self), field);
    }

    pub fn At(self: ?*anyopaque, ) i32 {
        return qtc.QSqlQuery_At(@ptrCast(self));
    }

    pub fn LastQuery(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSqlQuery_LastQuery(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsqlquery.LastQuery: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn NumRowsAffected(self: ?*anyopaque, ) i32 {
        return qtc.QSqlQuery_NumRowsAffected(@ptrCast(self));
    }

    pub fn LastError(self: ?*anyopaque, ) QtC.QSqlError {
        return qtc.QSqlQuery_LastError(@ptrCast(self));
    }

    pub fn IsSelect(self: ?*anyopaque, ) bool {
        return qtc.QSqlQuery_IsSelect(@ptrCast(self));
    }

    pub fn Size(self: ?*anyopaque, ) i32 {
        return qtc.QSqlQuery_Size(@ptrCast(self));
    }

    pub fn Driver(self: ?*anyopaque, ) QtC.QSqlDriver {
        return qtc.QSqlQuery_Driver(@ptrCast(self));
    }

    pub fn Result(self: ?*anyopaque, ) QtC.QSqlResult {
        return qtc.QSqlQuery_Result(@ptrCast(self));
    }

    pub fn IsForwardOnly(self: ?*anyopaque, ) bool {
        return qtc.QSqlQuery_IsForwardOnly(@ptrCast(self));
    }

    pub fn Record(self: ?*anyopaque, ) QtC.QSqlRecord {
        return qtc.QSqlQuery_Record(@ptrCast(self));
    }

    pub fn SetForwardOnly(self: ?*anyopaque, forward: bool) void {
        qtc.QSqlQuery_SetForwardOnly(@ptrCast(self), forward);
    }

    pub fn Exec(self: ?*anyopaque, query: []const u8) bool {
        const query_str = qtc.libqt_string{
    .len = query.len,
    .data = query.ptr,
};
return qtc.QSqlQuery_Exec(@ptrCast(self), query_str);
    }

    pub fn Value(self: ?*anyopaque, i: i32) QtC.QVariant {
        return qtc.QSqlQuery_Value(@ptrCast(self), i);
    }

    pub fn SetNumericalPrecisionPolicy(self: ?*anyopaque, precisionPolicy: i32) void {
        qtc.QSqlQuery_SetNumericalPrecisionPolicy(@ptrCast(self), @intCast(precisionPolicy));
    }

    pub fn NumericalPrecisionPolicy(self: ?*anyopaque, ) i32 {
        return qtc.QSqlQuery_NumericalPrecisionPolicy(@ptrCast(self));
    }

    pub fn SetPositionalBindingEnabled(self: ?*anyopaque, enable: bool) void {
        qtc.QSqlQuery_SetPositionalBindingEnabled(@ptrCast(self), enable);
    }

    pub fn IsPositionalBindingEnabled(self: ?*anyopaque, ) bool {
        return qtc.QSqlQuery_IsPositionalBindingEnabled(@ptrCast(self));
    }

    pub fn Seek(self: ?*anyopaque, i: i32) bool {
        return qtc.QSqlQuery_Seek(@ptrCast(self), i);
    }

    pub fn Next(self: ?*anyopaque, ) bool {
        return qtc.QSqlQuery_Next(@ptrCast(self));
    }

    pub fn Previous(self: ?*anyopaque, ) bool {
        return qtc.QSqlQuery_Previous(@ptrCast(self));
    }

    pub fn First(self: ?*anyopaque, ) bool {
        return qtc.QSqlQuery_First(@ptrCast(self));
    }

    pub fn Last(self: ?*anyopaque, ) bool {
        return qtc.QSqlQuery_Last(@ptrCast(self));
    }

    pub fn Clear(self: ?*anyopaque, ) void {
        qtc.QSqlQuery_Clear(@ptrCast(self));
    }

    pub fn Exec2(self: ?*anyopaque, ) bool {
        return qtc.QSqlQuery_Exec2(@ptrCast(self));
    }

    pub fn ExecBatch(self: ?*anyopaque, ) bool {
        return qtc.QSqlQuery_ExecBatch(@ptrCast(self));
    }

    pub fn Prepare(self: ?*anyopaque, query: []const u8) bool {
        const query_str = qtc.libqt_string{
    .len = query.len,
    .data = query.ptr,
};
return qtc.QSqlQuery_Prepare(@ptrCast(self), query_str);
    }

    pub fn BindValue(self: ?*anyopaque, placeholder: []const u8, val: ?*anyopaque) void {
        const placeholder_str = qtc.libqt_string{
    .len = placeholder.len,
    .data = placeholder.ptr,
};
qtc.QSqlQuery_BindValue(@ptrCast(self), placeholder_str, @ptrCast(val));
    }

    pub fn BindValue2(self: ?*anyopaque, pos: i32, val: ?*anyopaque) void {
        qtc.QSqlQuery_BindValue2(@ptrCast(self), pos, @ptrCast(val));
    }

    pub fn AddBindValue(self: ?*anyopaque, val: ?*anyopaque) void {
        qtc.QSqlQuery_AddBindValue(@ptrCast(self), @ptrCast(val));
    }

    pub fn BoundValue(self: ?*anyopaque, placeholder: []const u8) QtC.QVariant {
        const placeholder_str = qtc.libqt_string{
    .len = placeholder.len,
    .data = placeholder.ptr,
};
return qtc.QSqlQuery_BoundValue(@ptrCast(self), placeholder_str);
    }

    pub fn BoundValue2(self: ?*anyopaque, pos: i32) QtC.QVariant {
        return qtc.QSqlQuery_BoundValue2(@ptrCast(self), pos);
    }

    pub fn BoundValues(self: ?*anyopaque, ) []const u8 {
        return qtc.QSqlQuery_BoundValues(@ptrCast(self));
    }

    pub fn BoundValueNames(self: ?*anyopaque, ) []const u8 {
        return qtc.QSqlQuery_BoundValueNames(@ptrCast(self));
    }

    pub fn BoundValueName(self: ?*anyopaque, pos: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSqlQuery_BoundValueName(@ptrCast(self), pos);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsqlquery.BoundValueName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn ExecutedQuery(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSqlQuery_ExecutedQuery(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsqlquery.ExecutedQuery: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn LastInsertId(self: ?*anyopaque, ) QtC.QVariant {
        return qtc.QSqlQuery_LastInsertId(@ptrCast(self));
    }

    pub fn Finish(self: ?*anyopaque, ) void {
        qtc.QSqlQuery_Finish(@ptrCast(self));
    }

    pub fn NextResult(self: ?*anyopaque, ) bool {
        return qtc.QSqlQuery_NextResult(@ptrCast(self));
    }

    pub fn Seek2(self: ?*anyopaque, i: i32, relative: bool) bool {
        return qtc.QSqlQuery_Seek2(@ptrCast(self), i, relative);
    }

    pub fn ExecBatch1(self: ?*anyopaque, mode: i32) bool {
        return qtc.QSqlQuery_ExecBatch1(@ptrCast(self), @intCast(mode));
    }

    pub fn BindValue3(self: ?*anyopaque, placeholder: []const u8, val: ?*anyopaque, typeVal: i32) void {
        const placeholder_str = qtc.libqt_string{
    .len = placeholder.len,
    .data = placeholder.ptr,
};
qtc.QSqlQuery_BindValue3(@ptrCast(self), placeholder_str, @ptrCast(val), @intCast(typeVal));
    }

    pub fn BindValue32(self: ?*anyopaque, pos: i32, val: ?*anyopaque, typeVal: i32) void {
        qtc.QSqlQuery_BindValue32(@ptrCast(self), pos, @ptrCast(val), @intCast(typeVal));
    }

    pub fn AddBindValue2(self: ?*anyopaque, val: ?*anyopaque, typeVal: i32) void {
        qtc.QSqlQuery_AddBindValue2(@ptrCast(self), @ptrCast(val), @intCast(typeVal));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QSqlQuery_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/sqlquery.html#types
pub const enums = struct {
    pub const BatchExecutionMode = enum {
        pub const ValuesAsRows: i32 = 0;
        pub const ValuesAsColumns: i32 = 1;
    };

};
